include/odp/api/spec/ipsec.h

```diff
@@ -739,6 +739,36 @@ typedef struct odp_ipsec_sa_param_t {
 			 */
 			odp_cos_t dest_cos;
 
+			/** Enable reassembly of pre-IPsec IP fragments
+			 *
+			 *  When enabled, pktio input will attempt reassembly of
+			 *  pre-IPsec fragmented IP packets. This reassembly
+			 *  attempt would be performed following IPsec inbound
+			 *  processing.
+			 *
+			 *  The pktio input processing IPsec should have
+			 *  'ip_reassem' enabled to have this feature supported.
+			 *
+			 *  @see odp_pktio_config(), odp_reassembly_config()
+			 *
+			 */
+			odp_bool_t pre_frag_ip_reassem;
```

Comment:
Why does the fragment reassembly of the pktio need to be enabled? Why cannot the inner packet reassembly be independent of it? What happens if reassembly is not enabled at the pktio level?

What if I want reassembly for all packets coming in through IPsec tunnel (since I know they are for me) but not for any packets that are not IPsec decapsulated (since I may just forward them or since I do not want to open any potential DoS surface related to reassembly)?

There seems to be an underlyinh assumption that packets for the same SA come in through the same pktio, but that is not necessarily the case. In those cases the reference to _the_ pktio is a bit ambiguous.

How is failed reassembly handled? How the partial reassembly result is handed to the application in the IPsec case?

Reply:
> 
> 
> Why does the fragment reassembly of the pktio need to be enabled? Why cannot the inner packet reassembly be independent of it? What happens if reassembly is not enabled at the pktio level?

Disabling at the pktio level allows the application to not check reassembly status for received packets. Also, without pktio configuration, the hardware will need to have reassembly enabled by default which will limit the optimizations possible. If reassembly is disabled on the pktio interfaces on which inline is enabled, then the SA create will return error.

> 
> What if I want reassembly for all packets coming in through IPsec tunnel (since I know they are for me) but not for any packets that are not IPsec decapsulated (since I may just forward them or since I do not want to open any potential DoS surface related to reassembly)?

Good suggestion. I can introduce an extra flag to denote whether reassembly is supported in non-inline IPsec path. I can add the flag in capabilities and can allow the option to be controlled using config API. Shall I propose this as part of v3?

> 
> There seems to be an underlyinh assumption that packets for the same SA come in through the same pktio, but that is not necessarily the case. In those cases the reference to _the_ pktio is a bit ambiguous.

I'll rephrase as,
"The pktio inputs processing inline IPsec should have 'ip_reassem' enabled to have this feature supported."
Would that work?

> 
> How is failed reassembly handled? How the partial reassembly result is handed to the application in the IPsec case?

In case of "pre_frag_ip_reassem", reassembly would be attempted after IPsec decapsulation. In case of partial/incomplete reassembly, the application is expected to recover individual fragments using the proposed "odp_packet_recover_frags()" API. The packets would be identical to inline IPsec processed, and application is expected to begin afresh. The time already spend by the fragment waiting for reassembly will be available (during recover_frags()) to allow application to have accurate reassembly timeouts.

Reply:
> I'll rephrase as, "The pktio inputs processing inline IPsec should have 'ip_reassem' enabled to have this feature supported." Would that work?

I think it would be better to be more explicit on how the API works to remove any possible ambiguity. For instance, does the above mean that the feature must not be enabled if any of the pktio:s that have inline ipsec enabled have their reassembly flag cleared? Or is it ok to enable the feature in all cases but reassembly just may not happen if fragments arrive through pktio:s that do not have reass enabled? I would think it is the former, but it would be better if the API were more clear about it.